### PR TITLE
Hackathon 2025 contributions

### LINKED ISSUE #6: User configurable `max_new_tokens`
After `llmart` finds an adversarial suffix, the final response is obtained from the LLM by querying it with the final suffix.

Right now we hardcode `max_new_tokens` when the LLM responds:
https://github.com/IntelLabs/LLMart/blob/c7bbef39cd8958714a8d8c25654b1a0a20c6b1e4/src/llmart/attack.py#L381

Make this configurable by the end user since newer reasoning models will produce much longer responses.

### LINKED ISSUE #8: Customizable `bad_token_ids` policies
`llmart` has the capability of banning "bad" tokens from the adversarial optimization.

Right now `bad_token_ids` implements a static policy for what is considered a "bad" token (non-printability, ascii-only):
https://github.com/IntelLabs/LLMart/blob/c7bbef39cd8958714a8d8c25654b1a0a20c6b1e4/src/llmart/tokenizer.py#L428-L444

Being able to add configurable policies would help with non-ascii languages. Additionally, being able to ban a set of tokens would also be beneficial.

### LINKED ISSUE #12: Run test on attack with lowest loss
For some schedulers, like cosine, it possible for an attack at a prior step, and not the final step, to have a lower loss. We should keep track of the lowest train (or validation) loss attack and use that in the final step.

This is probably best done in conjunction with #11.

### LINKED ISSUE #30: Write documentation for `cfg`
We need documentation for all command line parameters. At least the type and one line for each.

### PR DESCRIPTION
**Pull request by mariusarvinte**
This PR includes several standalone contributions done during the hackathon:

- Add user configurable `max_new_tokens` for generation. Fixes #6
- Add customizable `bad_token_ids` policies. Fixes #8 
- Add config option to track `best_attack_token` based on the teacher forcing loss. Fixes #12
- Add CLI reference docs. Fixes #30

### CODE REVIEW

File: docs/cli-reference.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+| `experiment_name` | string | `llmart` | Name of the folder where results will be stored |
+| `output_dir` | string | `${now:%Y-%m-%d}/${now:%H-%M-%S.%f}` | Name of the sub-folder where results for the current run will be stored<br> Defaults to a millisecond-level timestamp |
+| `seed` | integer | 2024 | Global random seed for reproducibility<br>⚠️ The seed will only reproduce results on the same number of GPUs as the original run |
+| `use_deterministic_algorithms` | boolean | false | Whether to use cuDNN deterministic algorithms |
+| `steps` | integer | 500 | Number of adversarial optimization steps |
+| `early_stop` | boolean | true | Whether to enable early stopping<br> If `true`, enables early stopping once all forced tokens are rank-1 (guaranteed selection in greedy decoding) |
+| `val_every` | integer | 50 | Validation frequency (in steps) |
+| `save_every` | integer | 50 | Result saving frequency (in steps) |
+| `per_device_bs` | integer | 1 | Per-device batch size<br>⚠️ This parameter can greatly improve efficiency, but will error out if insufficient VRAM is available |
```
Comment (dxoigmn) [inline docs/cli-reference.md:35]: This can be `auto` soon...

File: src/llmart/attack.py [hunk 2]
```diff
@@ -210,11 +211,15 @@ def train(
 
     # For each optimization step
     step, results = 0, dict()
+    best_loss = float("inf")
+    best_attack_data = (
+        attack.module.param.data if hasattr(attack, "module") else attack.param.data
+    )
```
Comment (dxoigmn) [inline src/llmart/attack.py:217]: could we just deepcopy the `attack`? The proper way to do this is to use [`unwrap_model`](https://huggingface.co/docs/accelerate/v0.11.0/en/accelerator#accelerate.Accelerator.unwrap_model).

File: src/llmart/attack.py [hunk 3]
```diff
@@ -210,11 +211,15 @@ def train(
...
+
     for step, inputs in (
         pbar := tqdm(iterable=enumerate(train_dl), total=len(train_dl), desc="steps")
     ):
         optimizer.zero_grad()
-
```
Comment (dxoigmn) [inline src/llmart/attack.py:217]: not sure why this was removed

File: src/llmart/config.py [hunk 4]
```diff
@@ -448,6 +448,9 @@ class LLMartConf(CoreConf):
     early_stop: bool = True
     val_every: int = 50
     save_every: int = 50
+    keep_best_attack: bool = True
```
Comment (dxoigmn) [inline src/llmart/config.py:451]: I don't think this should be an option. We should always keep the attack with lowest loss.

File: src/llmart/attack.py [hunk 5]
```diff
@@ -285,6 +290,15 @@ def train(
             postfix["mem"] = f"{torch.cuda.max_memory_allocated()/(1024**2):0.3f}MiB"
             pbar.set_postfix(postfix)
 
+            # Save tokens with lowest training loss
+            if cfg.keep_best_attack and loss < best_loss:
+                attack_data = (
+                    attack.module.param.data
+                    if hasattr(attack, "module")
+                    else attack.param.data
+                )
```
Comment (dxoigmn) [inline src/llmart/attack.py:299]: deepcopy. at the very least this should be a function now that it is repeated.

File: src/llmart/tokenizer.py [hunk 6]
```diff
@@ -46,6 +46,7 @@ def __init__(
         self,
         tokenizer: PreTrainedTokenizerFast,
         tags: list[str] | None = None,
+        banned_strings: list[str] = [],
```
Comment (dxoigmn) [inline src/llmart/tokenizer.py:49]: mutable default args are not great: https://docs.python-guide.org/writing/gotchas/

**Review by dxoigmn — CHANGES_REQUESTED** [review]
Would be nice to be able to auto-generate `cli-reference.md` given we'll have changes to it once #64 is merged.

File: src/llmart/tokenizer.py (hunk 6 above)
Comment (mariusarvinte) [reply src/llmart/tokenizer.py]: Fixed

File: src/llmart/attack.py (hunk 2 above)
Comment (mariusarvinte) [reply src/llmart/attack.py]: Fixed and don't even need this, can just deepcopy the wrapped object

File: src/llmart/attack.py (hunk 3 above)
Comment (mariusarvinte) [reply src/llmart/attack.py]: Restored

File: src/llmart/config.py (hunk 4 above)
Comment (mariusarvinte) [reply src/llmart/config.py]: Fixed, but tracking highest train ASR instead

File: src/llmart/attack.py (hunk 5 above)
Comment (mariusarvinte) [reply src/llmart/attack.py]: Fixed

File: docs/cli-reference.md (hunk 1 above)
Comment (mariusarvinte) [reply docs/cli-reference.md]: Added

**Review by dxoigmn — APPROVED** [review]
LGTM!
